lib/logging/third_party_transaction.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module Logging
+  module ThirdPartyTransaction
+    module MethodWrapper
+      # @method_names
+      #   - Array (of /or single) method name(s). e.g. ['method_1', :method_2]
+      # @additional_logs
+      #   - Will add to and / or overwrite the default data with usecase specific
+      #     log data
+      def wrap_with_logging(*method_names, additional_logs: {})
+        # including the instance method helpers inside this method makes them
+        # available on the instance with access to the current scope, (e.g.
+        # current_user at the controller level)
+        include(ScopedInstanceMethods)
+
+        proxy = Module.new do
+          method_names.each do |method_name|
+            # define patchable method(s) with the same name inside of a proxy module.
+            define_method(method_name) do |*args, &block|
+              log_3pi_begin(method_name, additional_logs, *args)
+              # delegate to the original behavior
+              result = super(*args, &block)
+              log_3pi_complete(method_name, additional_logs, *args)
+              result
+            end
+          end
+        end
+        # prepend the proxy module, allowing both our wrapped version to run
+        # logging as well as delegating to the original.
+        prepend(proxy)
+      end
+    end
+
+    # these will be included after instance instantiation, making them available
+    # to the instance and retaining their scope.
+    module ScopedInstanceMethods
+      def log_3pi_begin(method_name, additional_logs = {}, args = [])
+        @start_time = Time.current
+
+        log = {
+          process_id: Process.pid,
+          user_uuid: try(:current_user).try(:account_uuid),
```

Comment:
1. Yes, I agree.  It would be cool to have this automagically scope itself to it's context, or exclude nil values from the log.  atm, i think this is fine, but yeah great note for a future refactor. 
2. `account_uuid` is  not necessarily universal, given (as you have seen) the weirdness or our User model.  I belive it may be dependent on how a user logs in.  I think another good future refactor could be changing this to `user_uuid: get_user_id` where `get_user_id`  tries a few values to return the salient user id 🤔 but... that's also a bit of weird indirection?  so I'd want to look into it before jumpting

tl;dr - good notes for future refactors imho